src/printer.rs

```diff
@@ -34,52 +41,92 @@ impl Printer {
         let theme = theme.unwrap_or(Theme::auto);
 
         Printer {
-            indent_json: matches!(pretty, Pretty::all | Pretty::format),
-            sort_headers: matches!(pretty, Pretty::all | Pretty::format),
-            color: matches!(pretty, Pretty::all | Pretty::colors),
-            stream: matches!(pretty, Pretty::none) || stream,
+            indent_json: pretty.format(),
+            sort_headers: pretty.format(),
+            color: pretty.color(),
+            stream,
             theme,
             buffer,
         }
     }
 
-    fn print_json(&mut self, text: &str) -> io::Result<()> {
-        // This code is a little thorny because of ownership issues.
-        // We have to keep the indented text alive until the end of the function.
-        let indent_result = match self.indent_json {
-            true => Some(indent_json(text)),
-            false => None,
-        };
-        let text = match &indent_result {
-            Some(Ok(result)) => result.as_str(),
-            _ => text,
-        };
-        if self.color {
-            colorize(text, "json", &self.theme, &mut self.buffer)
+    /// Run a piece of code with a [`Highlighter`] instance. After the code runs
+    /// successfully, [`Highlighter::finish`] will be called to properly terminate.
+    ///
+    /// That way you don't have to remember to call it manually, and errors
+    /// can still be handled (unlike an implementation of [`Drop`]).
+    fn with_highlighter(
+        &mut self,
+        syntax: &'static str,
+        code: impl FnOnce(&mut Highlighter) -> io::Result<()>,
+    ) -> io::Result<()> {
+        let mut highlighter = Highlighter::new(syntax, self.theme, &mut self.buffer);
+        code(&mut highlighter)?;
+        highlighter.finish()
+    }
+
+    fn colorize_text(&mut self, text: &str, syntax: &'static str) -> io::Result<()> {
+        self.with_highlighter(syntax, |highlighter| highlighter.highlight(text))
+    }
+
+    fn dump(&mut self, reader: &mut impl Read) -> io::Result<()> {
+        copy_largebuf(reader, &mut self.buffer)?;
+        Ok(())
+    }
+
+    fn print_json_text(&mut self, text: &str) -> io::Result<()> {
+        if !self.indent_json {
+            self.print_syntax_text(text, "json")
```

Comment:
So whether an output should colorized can be determined by either `print_json_text()` or `print_syntax_text`. I wonder if there is a way to restructure this so both indent and coloring are controlled by `print_json_text()`.

Reply:
I think that'd just mean duplicating `print_syntax_text` inside `print_json_text`, so it wouldn't really be cleaner.

Right now the logic is that if it doesn't have to do anything specialized it falls back to the generic version.

Reply:
Fair enough. Could you add your last sentence as a comment? I think it would help any future who is skimming the code.

Reply:
Thanks. Can we also add the same comment to `print_json_stream()`?

Reply:
Ah, sure.